### PR TITLE
One way to resolve a window along a patch's dimensions

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

One way to turn "a window of X" into sample counts. DASCore had four: `get_patch_window_size` (hampel, wiener), `_create_size_and_axes` in `proc/filter.py` (median, gaussian, savgol), `get_window_axis_step` (rolling, stft — the only one that understood percent, and one dimension only), and `AdaptiveSpectralFilter.geometry`. They differed at the edges — whether an even window is adjusted or refused, whether a window longer than the coordinate is refused, what `overlap=None` means — and every new windowed operation re-derived all of it.

`dascore.utils.window.resolve_window` replaces all four and returns a `Window` (dims, axes, size, overlap, stride, in samples). What differs between functions is policy, not conversion, so each policy is a keyword the function states: `require_odd`, `min_samples`, `warn_above`, `enforce_lt_coord`, `allow_multiple`, `allow_empty`, `default_overlap`. It takes a `PatchMeta` as readily as a patch, so a `PatchProcessor` can resolve a window without data, and it accepts overlap as a scalar or a per-dimension mapping — which only the adaptive spectral filter uses yet.

This is the first PR of the windowing plan (top-level `.scratch/windowing-unification-plan.md`), and it is meant to change nothing a user can see. `tests/test_utils/test_window_parity.py` pins the edges of every migrated function with literal values taken from `dev` before the change — hampel rounding 4 samples up to 5 in units but refusing 4 in samples, median accepting even windows, rolling and stft refusing oversize windows while the dense filters pass them to scipy, stft's 50% of an odd window rounding half-to-even (5 → hop 3, 7 → 3, 9 → 5), a quantity keeping its units under `samples=True`, and savgol keeping only the last keyword's axis (a bug; pinned, not fixed — see below). Those tests passed on `dev` before the migration and pass after it.

What is deliberately not identical — each a call which used to fail and now works, pinned in `TestChangedOnPurpose`:

- A `Quantity` window under `samples=True` used to raise for the dense filters (`get_sample_count` refused it) while rolling and stft accepted it. Now a quantity carries its own units everywhere, and hampel rounds it up to odd as it would a value in units.
- `overlap={"time": 2}` used to raise a `TypeError` for rolling and stft; a mapping is now one value per dimension for every function.
- The adaptive spectral filter read `overlap=25 * percent` as `int(0.25) == 0` and filtered with no overlap; it now reads 25% of the window. `overlap={"time": None}` raised a `TypeError` and now says to leave the key out.
- `get_patch_window_size` and `get_window_axis_step` remain in `dascore.utils.patch` as deprecated wrappers over `resolve_window`, removed in 0.2.0.
- `stft(time=0, samples=True)` used to reach scipy and fail on the window's shape; it now raises the resolver's "at least 1 samples" `ParameterError`.
- Several adaptive spectral filter messages changed wording (the zero-window, negative-overlap, and unknown-dimension ones); they landed yesterday in #670.

Not migrated, on purpose: `kurtosis` (its own floor of 2 and a sorted-coordinate rule, with a message tests match), `whiten` (a smoothing length on a frequency coordinate, not a window), and `viz.spectrogram` (calls legacy `scipy.signal.spectrogram` with raw kwargs). Each is noted in the plan.

### A bug this PR pins rather than fixes

`savgol_filter` over two dimensions filters `patch.data` on every pass instead of the previous pass's output, so only the last keyword's axis survives: `savgol_filter(time=5, distance=7)` equals `savgol_filter(distance=7)`. The parity test documents it; the fix is one word (`x=data`) and its own PR, because it changes results.

### Codex review

Requested on the plan before this PR and on the code after it; both under `.scratch/`, untracked. The code review found the first commit less faithful than it claimed — two P1s (AFK on uneven coordinates with `samples=True` regressed to an error; its percent overlap changed from a silent 0) and five P2/P3s (dense-filter floors, oddness judged by the wrong flag, defaults unchecked, check precedence, the deleted utilities). All are fixed in the second commit and each has a pinned test. A third, cleanliness-focused review found a step longer than the window (which rolling allows) surfacing as a negative `overlap` in a field documented non-negative; `Window` now stores the stride and derives the overlap, negative across a gap and documented so. The plan review's P1s shaped this PR: per-call policies rather than a fixed signature, `overlap=None` left as None for the caller to read (rolling → stride 1, stft → windows abut, AFK → `window // 2 - 1`), stft's half-to-even rounding specified and pinned, no new public arguments, and characterization tests written first.

## Changelog

- none

## Checklist

I have:

- [x] filled in the Changelog section above (see `docs/contributing/general_guidelines.qmd`).

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [x] added the "ready_for_review" tag once the PR is ready to be reviewed.

## Summary by CodeRabbit

* **New Features**
  * Added consistent window and overlap handling across filters, rolling operations, and Fourier transforms.
  * Added support for dimension-aware windows, sample- or coordinate-based sizing, percentages, strides, and overlaps.
  * Added centralized validation for window sizes, sampling, dimensions, and coordinate limits.

* **Bug Fixes**
  * Improved handling of unevenly sampled coordinates and zero-sized windows.
  * Standardized validation and error reporting across windowed operations.

* **Tests**
  * Expanded coverage for window resolution, overlap, stride, validation, and edge cases.

### CODE REVIEW

File: dascore/utils/patch.py [hunk 1]
```diff
@@ -928,196 +926,6 @@ def get_dim_sampling_rate(patch: PatchType, dim: str) -> float:
     return 1.0 / d_dim
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/utils/patch.py:937]: Kept in 4f86578e: both remain in dascore.utils.patch as @deprecate wrappers over resolve_window (removed_in="0.2.0"), with tests in TestDeprecatedResolvers asserting the same answers plus the warning.

File: dascore/utils/window.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    sizes = []
+    for dim, _, value in dim_axis_values:
+        coord = coords[dim]
+        count = _to_samples(
+            coord, value, samples=samples, enforce_lt_coord=enforce_lt_coord
+        )
+        if count < min_samples:
+            raise ParameterError(_too_small(dim, coord, count, min_samples, samples))
+        if require_odd and count % 2 != 1:
+            if samples:
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/utils/window.py]: Fixed in 4f86578e: _to_samples now returns whether the value was read as a sample count, and the odd-window branch judges by that rather than the samples flag, so hampel_filter(time=16 * ms, samples=True) rounds 4 up to 5. Pinned in TestPolicies.test_quantity_under_samples_adjusts_to_odd and TestChangedOnPurpose.test_hampel_takes_a_quantity_under_samples.

File: dascore/proc/adaptive_spectral_filter.py [hunk 3]
```diff
@@ -504,16 +444,22 @@ def geometry(self, meta: PatchMeta) -> _Geometry:
...
-        dims = tuple(selected)
-        axes = _axes(meta, dims)
-        windows = _sample_counts(meta, selected, samples=self.samples, name="window")
-        overlap_values, sample_dims = _normalize_overlap(self.overlap, dims, windows)
-        overlaps = _sample_counts(
+        window = resolve_window(
             meta,
-            overlap_values,
+            selected,
             samples=self.samples,
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/adaptive_spectral_filter.py:450]: Fixed in 4f86578e: require_evenly_sampled=False now also means a bare sample count is read as given without consulting the coordinate, which is what AFK passes. wacky_dim_coords_patch.adaptive_spectral_filter(time=16, samples=True) works again; pinned in TestUnevenCoordinatesInSamples, alongside the dense filters and rolling which refused before and still do.

File: dascore/utils/window.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    given = _spread(step if step is not None else overlap, dims, name)
+    out: list[int | None] = []
+    for dim, window in zip(dims, size):
+        value = given[dim]
+        if value is None:
+            if default is None:
+                out.append(None)
+                continue
+            out.append(default if isinstance(default, int) else default(window))
+            continue
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/utils/window.py]: Fixed in 4f86578e: a default goes through the same non-negative and positive-advance checks a given overlap does. default_overlap=window_size now raises "Window step must be greater than zero"; pinned in TestPolicies.test_default_overlap_is_checked.

File: dascore/utils/window.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        if count < min_samples:
+            raise ParameterError(_too_small(dim, coord, count, min_samples, samples))
+        if require_odd and count % 2 != 1:
+            if samples:
+                msg = (
+                    f"For clean median calculation, dimension windows must be odd "
+                    f"but {dim} has a value of {count} samples."
+                )
+                raise ParameterError(msg)
+            count += 1
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/utils/window.py]: Fixed in 4f86578e — same root cause as the codex thread above: oddness is now judged by whether the value turned out to be a sample count, not by the samples flag. Test: TestPolicies.test_quantity_under_samples_adjusts_to_odd.
